File: java_projects/glue_code/commons-validator/src/main/java/org/apache/commons/validator/routines/checkdigit/IBANCheckDigit.java
package org.apache.commons.validator.routines.checkdigit;

import java.io.Serializable;
import org.apache.commons.validator.ContextInitializer;
import org.apache.commons.validator.ExceptionHandler;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

public final class IBANCheckDigit implements CheckDigit, Serializable {
  public static final CheckDigit IBAN_CHECK_DIGIT = new IBANCheckDigit();
  private static final long MODULUS = 97;
  private static final long MAX = 999999999;
  private static final int MAX_ALPHANUMERIC_VALUE = 35; // Character.getNumericValue('Z')
  private static final long serialVersionUID = -3600191725934382801L;
  private static final int MIN_CODE_LEN = 5;
  private static Value clz =
      ContextInitializer.getPythonClass("/routines/checkdigit/IBANCheckDigit.py", "IBANCheckDigit");
  private Value obj;

  public IBANCheckDigit(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public String calculate(String code) throws CheckDigitException {
    try {
      //
      // if (code == null || code.length() < MIN_CODE_LEN) {
      // throw CheckDigitException.CheckDigitException1(
      // "Invalid Code length=" + (code == null ? 0 : code.length()));
      // }
      // code = code.substring(0, 2) + "00" + code.substring(4); // CHECKSTYLE IGNORE MagicNumber
      // int modulusResult = calculateModulus(code);
      // int charValue = (98 - modulusResult); // CHECKSTYLE IGNORE MagicNumber
      // String checkDigit = Integer.toString(charValue);
      // return (charValue > 9 ? checkDigit : "0" + checkDigit); // CHECKSTYLE IGNORE MagicNumber
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("calculate", code).as(String.class);
    } catch (PolyglotException e) {
      // TODO: Handle CheckDigitException
      throw (CheckDigitException) ExceptionHandler.handle(e, "IBANCheckDigit.calculate");
    }
  }

  public boolean isValid(String code) {
    //
    // if (code == null || code.length() < MIN_CODE_LEN) {
    // return false;
    // }
    // String check = code.substring(2, 4); // CHECKSTYLE IGNORE MagicNumber
    // if ("00".equals(check) || "01".equals(check) || "99".equals(check)) {
    // return false;
    // }
    // try {
    // int modulusResult = calculateModulus(code);
    // return (modulusResult == 1);
    // } catch (CheckDigitException ex) {
    // return false;
    // }
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isValid", code).as(boolean.class);
  }

  public IBANCheckDigit() {
    //

    this.obj = clz.invokeMember("__init__");
  }

  private int calculateModulus(String code) throws CheckDigitException {
    try {
      //
      // String reformattedCode =
      // code.substring(4) + code.substring(0, 4); // CHECKSTYLE IGNORE MagicNumber
      // long total = 0;
      // for (int i = 0; i < reformattedCode.length(); i++) {
      // int charValue = Character.getNumericValue(reformattedCode.charAt(i));
      // if (charValue < 0 || charValue > MAX_ALPHANUMERIC_VALUE) {
      // throw CheckDigitException.CheckDigitException1(
      // "Invalid Character[" + i + "] = '" + charValue + "'");
      // }
      // total =
      // (charValue > 9 ? total * 100 : total * 10)
      // + charValue; // CHECKSTYLE IGNORE MagicNumber
      // if (total > MAX) {
      // total = total % MODULUS;
      // }
      // }
      // return (int) (total % MODULUS);
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("calculateModulus", code).as(int.class);
    } catch (PolyglotException e) {
      // TODO: Handle CheckDigitException
      throw (CheckDigitException) ExceptionHandler.handle(e, "IBANCheckDigit.calculateModulus");
    }
  }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/body/commons-validator/src/main/org/apache/commons/validator/routines/checkdigit/IBANCheckDigit.py
from __future__ import annotations
import re
import io
from src.main.org.apache.commons.validator.routines.checkdigit.CheckDigit import *
from src.main.org.apache.commons.validator.routines.checkdigit.CheckDigitException import *


class IBANCheckDigit(CheckDigit):

    __MODULUS: int = 97
    __MAX: int = 999999999
    __MAX_ALPHANUMERIC_VALUE: int = 35  # Character.getNumericValue('Z')
    __serialVersionUID: int = -3600191725934382801
    __MIN_CODE_LEN: int = 5

    def calculate(self, code: str) -> str:

        if code is None or len(code) < self.__MIN_CODE_LEN:
            raise CheckDigitException.CheckDigitException1(
                "Invalid Code length=" + (str(0) if code is None else str(len(code)))
            )
        code = code[:2] + "00" + code[4:]
        modulusResult = self.__calculateModulus(code)
        charValue = 98 - modulusResult
        checkDigit = str(charValue)
        return checkDigit if charValue > 9 else "0" + checkDigit

    def isValid(self, code: str) -> bool:

        if code is None or len(code) < self.__MIN_CODE_LEN:
            return False
        check = code[2:4]
        if check == "00" or check == "01" or check == "99":
            return False
        try:
            modulusResult = self.__calculateModulus(code)
            return modulusResult == 1
        except CheckDigitException:
            return False

    def __init__(self) -> None:

        self.__MODULUS: int = 97
        self.__MAX: int = 999999999
        self.__MAX_ALPHANUMERIC_VALUE: int = 35  # Character.getNumericValue('Z')
        self.__serialVersionUID: int = -3600191725934382801
        self.__MIN_CODE_LEN: int = 5
        self.IBAN_CHECK_DIGIT: CheckDigit = CheckDigit()

    def __calculateModulus(self, code: str) -> int:

        reformattedCode = code[4:] + code[:4]
        total = 0
        for i in range(len(reformattedCode)):
            charValue = (
                int(reformattedCode[i])
                if reformattedCode[i].isdigit()
                else ord(reformattedCode[i]) - 55
            )
            if charValue < 0 or charValue > self.__MAX_ALPHANUMERIC_VALUE:
                raise CheckDigitException.CheckDigitException1(
                    "Invalid Character[" + str(i) + "] = '" + str(charValue) + "'"
                )
            total = (total * 100 if charValue > 9 else total * 10) + charValue
            if total > self.__MAX:
                total = total % self.__MODULUS
        return total % self.__MODULUS
